src/common/random_generator.h

```diff
@@ -0,0 +1,224 @@
+    GType dist_uniform;
+    return dist_uniform(*engine_);
+  }
+
+  MSHADOW_XINLINE FType normal() {
+    std::normal_distribution<FType> dist_normal;
+    return dist_normal(*engine_);
+  }
+
+ private:
+  std::mt19937 *engine_;
+};
+
+template<typename DType>
+class RandGenerator<cpu, DType> {
+ public:
+  MSHADOW_XINLINE void dispose() {}
+
+  MSHADOW_XINLINE RandGeneratorImpl<cpu, DType> Get(int idx = 0) {
+    std::mt19937 *ptr_engine = &states_[idx];
+    RandGeneratorImpl<cpu, DType> gen(ptr_engine);
+    return gen;
+  }
+
+  MSHADOW_XINLINE void Seed(Stream<cpu> *, uint32_t seed) {
+    for (int i = 0; i < kCPURndStateNum; ++i) (states_ + i)->seed(seed + i);
+  }
+
+  MSHADOW_XINLINE void set_state(int idx, const std::mt19937 &state) {
+    states_[idx] = state;
+  }
+
+ private:
+  std::mt19937 states_[kCPURndStateNum];
+};
+
+#if MXNET_USE_CUDA
+
+// at least how many random numbers should be generated by one GPU thread.
+const int kGPUMinRndNumberPerThread = 64;
+// store how many global random states for GPU.
+const int kGPURndStateNum = 32768;
+
+// uniform number generation in Cuda made consistent with stl (include 0 but exclude 1)
+// by using 1.0-curand_uniform().
+// Needed as some samplers in sampler.h won't be able to deal with
+// one of the boundary cases.
+template<typename DType>
+class RandGeneratorImpl<gpu, DType> {
+ public:
+  // Copy state to local memory for efficiency.
+  __device__ explicit RandGeneratorImpl(curandStatePhilox4_32_10_t *state)
+      : state_(*state) {}
+
+  MSHADOW_FORCE_INLINE __device__ int rand() {
+    return curand(&state_);
+  }
+
+  MSHADOW_FORCE_INLINE __device__ float uniform() {
+    return static_cast<float>(1.0) - curand_uniform(&state_);
+  }
+
+  MSHADOW_FORCE_INLINE __device__ float normal() {
+    return curand_normal(&state_);
+  }
+
+  MSHADOW_FORCE_INLINE __device__ curandStatePhilox4_32_10_t get_state() {
+    return state_;
+  }
+
+ private:
+  curandStatePhilox4_32_10_t state_;
+};
+
+template<>
+class RandGeneratorImpl<gpu, double> {
+ public:
+  // Copy state to local memory for efficiency.
+  __device__ explicit RandGeneratorImpl(curandStatePhilox4_32_10_t *state)
+      : state_(*state) {}
+
+  MSHADOW_FORCE_INLINE __device__ int rand() {
+    return curand(&state_);
+  }
+
+  MSHADOW_FORCE_INLINE __device__ double uniform() {
+    return static_cast<double>(1.0) - curand_uniform_double(&state_);
+  }
+
+  MSHADOW_FORCE_INLINE __device__ double normal() {
+    return curand_normal_double(&state_);
+  }
+
+  MSHADOW_FORCE_INLINE __device__ curandStatePhilox4_32_10_t get_state() {
+    return state_;
+  }
+
+ private:
+  curandStatePhilox4_32_10_t state_;
+};
+
+template<typename DType>
+class RandGenerator<gpu, DType> {
+ public:
+  RandGenerator() {
+    cudaError_t e = cudaMalloc(&states_, kGPURndStateNum * sizeof(curandStatePhilox4_32_10_t));
+    if (e != cudaSuccess && e != cudaErrorCudartUnloading) {
+      throw std::bad_alloc();
```

Comment:
Why not using the existing macros for interpreting cuda-errors (i.e. CUDA_CALL(cudaMalloc(.....))? That would also tell the user that something went wrong on the device, while throwing std::bad_alloc will be misleading as it refers to a memory allocation on the host.

Reply:
fixed.